File: CrossAttention.py
class CrossAttention(nn.Module):

    def __init__(self, embed_dim, num_heads):
        super(CrossAttention, self).__init__()
        self.attention = nn.MultiheadAttention(embed_dim, num_heads)
        #self.tensor1 = tensor1
        #self.tensor2 = tensor2


    def forward(self, queries, key, values):
        batch_size, _, depth, heaight, width = queries.size()
        queries = queries.permute(2, 0, 1, 3, 4).reshape(depth, -1, queries.size(-1))
        key = key.permute(2, 0, 1, 3, 4).reshape(depth, -1, key.size(-1))
        values = values.permute(2, 0, 1, 3, 4).reshape(depth, -1, values.size(-1))
        output, _ = self.attention(queries, key, values)

        output = output.view(depth, batch_size, -1, queries.size(-1), queries.size(-1))
        #print('output shape: ', output.shape)
        output = output.permute(1, 2, 0, 3, 4)
        print('output shape: ', output.shape)
        return output

File: CrossAttention.cpp

#include <torch/torch.h>
#include <torch/script.h>
#include <ATen/cuda/CUDAContext.h>

class CrossAttention : public torch::nn::Module {
public:
    CrossAttention(int64_t embed_dim, int64_t num_heads) {
        attention = register_module("attention", torch::nn::MultiheadAttention(embed_dim, num_heads));
    }

    torch::Tensor forward(torch::Tensor queries, torch::Tensor key, torch::Tensor values) {
        int64_t batch_size = queries.size(0);
        int64_t depth = queries.size(1);
        int64_t height = queries.size(2);
        int64_t width = queries.size(3);

        queries = queries.permute({1, 0, 2, 3}).reshape({depth, -1, queries.size(-1)});
        key = key.permute({1, 0, 2, 3}).reshape({depth, -1, key.size(-1)});
        values = values.permute({1, 0, 2, 3}).reshape({depth, -1, values.size(-1)});

        torch::Tensor output;
        AT_DISPATCH_FLOATING_TYPES(queries.scalar_type(), "CrossAttention forward", ([&] {
            output = attention->forward(queries, key, values).toTuple()->get<0>();
        }));

        output = output.view({depth, batch_size, -1, queries.size(-1), queries.size(-1)});
        output = output.permute({1, 2, 0, 3, 4});

        return output;
    }

private:
    torch::nn::MultiheadAttention attention;
};
